syft/source/filesource/file_source.go

```diff
@@ -246,15 +239,69 @@ func digestOfFileContents(path string) string {
 	return di.String()
 }
 
-func unarchiveToTmp(path string, unarchiver archiver.Unarchiver) (string, func() error, error) {
+func unarchiveToTmp(path string, unarchiver archives.Extractor) (string, func() error, error) {
+	var cleanupFn = func() error { return nil }
+	archive, err := os.Open(path)
+	if err != nil {
+		return "", cleanupFn, fmt.Errorf("unable to open archive: %v", err)
+	}
+	defer archive.Close()
+
 	tempDir, err := os.MkdirTemp("", "syft-archive-contents-")
 	if err != nil {
-		return "", func() error { return nil }, fmt.Errorf("unable to create tempdir for archive processing: %w", err)
+		return "", cleanupFn, fmt.Errorf("unable to create tempdir for archive processing: %w", err)
 	}
 
-	cleanupFn := func() error {
-		return os.RemoveAll(tempDir)
+	visitor := func(_ context.Context, file archives.FileInfo) error {
+		// Protect against symlink attacks by ensuring path doesn't escape tempDir
+		destPath, err := safeJoinPath(tempDir, file.NameInArchive)
+		if err != nil {
+			return err
+		}
+
+		if file.IsDir() {
+			return os.MkdirAll(destPath, file.Mode())
+		}
+
+		if err = os.MkdirAll(filepath.Dir(destPath), os.ModeDir|0755); err != nil {
+			return fmt.Errorf("failed to create parent directory: %w", err)
+		}
+
+		rc, err := file.Open()
+		if err != nil {
+			return fmt.Errorf("failed to open file in archive: %w", err)
+		}
+		defer rc.Close()
+
+		destFile, err := os.Create(destPath)
+		if err != nil {
+			return fmt.Errorf("failed to create file in destination: %w", err)
+		}
+		defer destFile.Close()
+
+		if err := destFile.Chmod(file.Mode()); err != nil {
+			return fmt.Errorf("failed to change mode of destination file: %w", err)
+		}
+
+		if _, err := io.Copy(destFile, rc); err != nil {
+			return fmt.Errorf("failed to copy file contents: %w", err)
+		}
+
+		return nil
 	}
 
-	return tempDir, cleanupFn, unarchiver.Unarchive(path, tempDir)
+	return tempDir, func() error {
+		return os.RemoveAll(tempDir)
+	}, unarchiver.Extract(context.Background(), archive, visitor)
+}
+
+// safeJoinPath ensures that any destinations do not resolve to a path above the prefix path.
+// This protects against directory traversal attacks (zip slip).
+func safeJoinPath(prefix string, dest ...string) (string, error) {
```

Comment:
this is already implemented in `zip_file_traversal.go` -- can we reuse this instead of reimplementing?

Reply:
reused the internal `SafeJoin` in `filesource` now